core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java

```diff
@@ -54,6 +54,8 @@
 
     private WriteConsistencyLevel consistencyLevel = WriteConsistencyLevel.DEFAULT;
 
+    private long minimumClusterStateVersion = 0;
```

Comment:
I see where you are going with the minimumClusterStateVersion here. I would prefer being more explicit and naming it `routedBasedOnClusterStateVersion`  or something that implies when it's set.  In your original suggestion this was a hard requirement, but now we only use it on failure to resolve a primary.